File: freeq-web2/lib/session_store.rb
# frozen_string_literal: true

require "openssl"
require "json"
require "fileutils"
require "securerandom"
require "pathname"

require_relative "atproto/o_auth_session"
require_relative "atproto/dpop_key"

# Disk-backed encrypted OAuth session store.
#
# Mirrors freeq-webui's SessionStore (src/state.rs):
#   - one file per browser session_id under a local directory
#   - AES-256-GCM with a per-session key derived via HKDF-SHA256 from a
#     machine-local secret + the session_id
#   - file format: nonce (12) || ciphertext || tag (16)
#
# Env: FREEQ_WEB2_SESSIONS_DIR — default `.dev-data/web2-sessions`.
#      Set empty to disable disk persistence.
class SessionStore
  INFO = "freeq-session-encryption"
  NONCE_LEN = 12
  TAG_LEN = 16
  KEY_LEN = 32

  attr_reader :dir

  # Open or create the store. Returns nil when persistence is disabled.
  def self.open
    raw = ENV.fetch("FREEQ_WEB2_SESSIONS_DIR", ".dev-data/web2-sessions")
    return nil if raw.to_s.empty?

    dir = Pathname.new(raw)
    dir = Rails.root.join(dir) if defined?(Rails) && !dir.absolute?
    new(dir)
  rescue StandardError => e
    warn_log("session store init failed: #{e.class}: #{e.message}")
    nil
  end

  def initialize(dir)
    @dir = Pathname(dir)
    FileUtils.mkdir_p(@dir)
    @machine_key = load_or_create_machine_key
  end

  def save(sid, oauth)
    return if sid.to_s.empty? || oauth.nil?

    path = session_path(sid)
    key = derive_key(sid)
    plaintext = JSON.generate(payload(oauth))
    blob = encrypt(plaintext, key)
    atomic_write(path, blob)
  end

  def load(sid)
    return nil if sid.to_s.empty?

    path = session_path(sid)
    return nil unless path.exist?

    key = derive_key(sid)
    plaintext = decrypt(path.binread, key)
    data = JSON.parse(plaintext)
    Atproto::OAuthSession.from_h(data)
  rescue StandardError => e
    warn_log("load session #{sid[0, 8]}… failed: #{e.class}: #{e.message}")
    nil
  end

  def remove(sid)
    return if sid.to_s.empty?

    path = session_path(sid)
    path.delete if path.exist?
    channels_path(sid).delete if channels_path(sid).exist?
  rescue StandardError => e
    warn_log("remove session #{sid[0, 8]}… failed: #{e.class}: #{e.message}")
  end

  # ── Channel list (client-authoritative) ────────────────────────────
  #
  # freeq-web2 owns the user's joined-channel list — the upstream server
  # is treated as a dumb relay. Persisted (encrypted like the OAuth blob)
  # so the list survives process restarts and can be re-asserted on every
  # fresh WS connect.

  def save_channels(sid, channels)
    return if sid.to_s.empty?

    key = derive_key(sid)
    plaintext = JSON.generate(channels.map(&:to_s))
    atomic_write(channels_path(sid), encrypt(plaintext, key))
  rescue StandardError => e
    warn_log("save channels #{sid[0, 8]}… failed: #{e.class}: #{e.message}")
  end

  def load_channels(sid)
    return [] if sid.to_s.empty?

    path = channels_path(sid)
    return [] unless path.exist?

    key = derive_key(sid)
    data = JSON.parse(decrypt(path.binread, key))
    data.is_a?(Array) ? data.map(&:to_s) : []
  rescue StandardError => e
    warn_log("load channels #{sid[0, 8]}… failed: #{e.class}: #{e.message}")
    []
  end

  private

  def session_path(sid)
    safe = sid.to_s.gsub(%r{[/\\.]}, "_")
    @dir.join("#{safe}.bin")
  end

  def channels_path(sid)
    safe = sid.to_s.gsub(%r{[/\\.]}, "_")
    @dir.join("#{safe}.channels")
  end

  def load_or_create_machine_key
    key_path = @dir.join(".key")
    if key_path.exist?
      bytes = key_path.binread
      raise "session key file has wrong length: #{bytes.bytesize}" unless bytes.bytesize == KEY_LEN

      return bytes
    end

    key = SecureRandom.random_bytes(KEY_LEN)
    atomic_write(key_path, key)
    key
  end

  # HKDF-SHA256(ikm=machine_key, salt=sid, info="freeq-session-encryption") → 32B
  # Matches freeq_sdk::oauth::derive_session_key.
  def derive_key(sid)
    OpenSSL::KDF.hkdf(
      @machine_key,
      salt: sid.to_s,
      info: INFO,
      length: KEY_LEN,
      hash: "SHA256"
    )
  end

  # Must include refresh fields — without them restored sessions skip
  # refresh_oauth_before_sasl! and SASL 904s once the access token expires.
  def payload(oauth)
    {
      "did" => oauth.did,
      "handle" => oauth.handle,
      "access_token" => oauth.access_token,
      "pds_url" => oauth.pds_url,
      "dpop_key" => oauth.dpop_key.serialize,
      "dpop_nonce" => oauth.dpop_nonce,
      "refresh_token" => oauth.refresh_token,
      "token_endpoint" => oauth.token_endpoint,
      "client_id" => oauth.client_id
    }
  end

  def encrypt(plaintext, key)
    cipher = OpenSSL::Cipher.new("aes-256-gcm")
    cipher.encrypt
    cipher.key = key
    nonce = cipher.random_iv
    raise "unexpected GCM iv length #{nonce.bytesize}" unless nonce.bytesize == NONCE_LEN

    ciphertext = cipher.update(plaintext) + cipher.final
    tag = cipher.auth_tag(TAG_LEN)
    nonce + ciphertext + tag
  end

  def decrypt(data, key)
    raise "encrypted session file too short" if data.bytesize < NONCE_LEN + TAG_LEN

    nonce = data.byteslice(0, NONCE_LEN)
    tag = data.byteslice(-TAG_LEN, TAG_LEN)
    ciphertext = data.byteslice(NONCE_LEN...-TAG_LEN)

    cipher = OpenSSL::Cipher.new("aes-256-gcm")
    cipher.decrypt
    cipher.key = key
    cipher.iv = nonce
    cipher.auth_tag = tag
    cipher.update(ciphertext) + cipher.final
  end

  def atomic_write(path, bytes)
    tmp = path.sub_ext("#{path.extname}.#{Process.pid}.tmp")
    File.binwrite(tmp, bytes)
    File.chmod(0o600, tmp)
    File.rename(tmp, path)
    File.chmod(0o600, path)
  ensure
    File.delete(tmp) if tmp && File.exist?(tmp)
  end

  def self.warn_log(msg)
    if defined?(Rails) && Rails.respond_to?(:logger) && Rails.logger
      Rails.logger.warn(msg)
    else
      warn("[SessionStore] #{msg}")
    end
  end

  def warn_log(msg)
    self.class.warn_log(msg)
  end
end

File: freeq-web3/lib/freeq_web3/session_store.ex
defmodule FreeqWeb3.SessionStore do
  @moduledoc """
  Disk-backed encrypted OAuth session store.

  Mirrors freeq-web2 `SessionStore` / freeq-webui:
  - one file per browser `session_id` under a local directory
  - AES-256-GCM with a per-session key derived via HKDF-SHA256 from a
    machine-local secret + the session_id
  - file format: nonce (12) || ciphertext || tag (16)

  Env: `FREEQ_WEB3_SESSIONS_DIR` — default `.dev-data/web3-sessions`.
  Set empty to disable disk persistence.
  """

  require Logger

  alias FreeqWeb3.Atproto.OAuthSession

  @info "freeq-session-encryption"
  @nonce_len 12
  @tag_len 16
  @key_len 32

  defstruct [:dir, :machine_key]

  @type t :: %__MODULE__{dir: String.t(), machine_key: binary()}

  @doc """
  Open or create the store. Returns `nil` when persistence is disabled
  (empty dir config) or initialization fails.
  """
  def open do
    raw =
      Application.get_env(
        :freeq_web3,
        :sessions_dir,
        System.get_env("FREEQ_WEB3_SESSIONS_DIR", ".dev-data/web3-sessions")
      )

    if raw in [nil, ""] do
      nil
    else
      dir = expand_dir(raw)
      File.mkdir_p!(dir)
      machine_key = load_or_create_machine_key(dir)
      %__MODULE__{dir: dir, machine_key: machine_key}
    end
  rescue
    e ->
      Logger.warning("session store init failed: #{Exception.message(e)}")
      nil
  end

  @doc "Save OAuth credentials for `session_id`. No-op when store disabled."
  def save(sid, %OAuthSession{} = oauth) when is_binary(sid) do
    if sid == "" do
      :ok
    else
      with %__MODULE__{} = store <- open_cached() do
        path = session_path(store, sid)
        key = derive_key(store, sid)
        plaintext = Jason.encode!(OAuthSession.to_map(oauth))
        blob = encrypt(plaintext, key)
        atomic_write!(path, blob)
      end

      :ok
    end
  rescue
    e ->
      Logger.warning("save session #{short(sid)} failed: #{Exception.message(e)}")
      :ok
  end

  def save(_, _), do: :ok

  @doc "Load OAuth credentials. Returns `%OAuthSession{}` or `nil`."
  def load(sid) when is_binary(sid) do
    if sid == "" do
      nil
    else
      with %__MODULE__{} = store <- open_cached(),
           path = session_path(store, sid),
           true <- File.exists?(path),
           key = derive_key(store, sid),
           {:ok, raw} <- File.read(path),
           plaintext when is_binary(plaintext) <- decrypt(raw, key),
           {:ok, data} <- Jason.decode(plaintext) do
        OAuthSession.from_map(data)
      else
        _ -> nil
      end
    end
  rescue
    e ->
      Logger.warning("load session #{short(sid)} failed: #{Exception.message(e)}")
      nil
  end

  def load(_), do: nil

  @doc "Drop persisted OAuth + channel list for logout."
  def remove(sid) when is_binary(sid) do
    if sid != "" do
      with %__MODULE__{} = store <- open_cached() do
        path = session_path(store, sid)
        ch_path = channels_path(store, sid)
        if File.exists?(path), do: File.rm(path)
        if File.exists?(ch_path), do: File.rm(ch_path)
      end
    end

    :ok
  rescue
    e ->
      Logger.warning("remove session #{short(sid)} failed: #{Exception.message(e)}")
      :ok
  end

  def remove(_), do: :ok

  @doc "Persist client-authoritative channel list."
  def save_channels(sid, channels) when is_binary(sid) and is_list(channels) do
    if sid == "" do
      :ok
    else
      with %__MODULE__{} = store <- open_cached() do
        key = derive_key(store, sid)
        plaintext = Jason.encode!(Enum.map(channels, &to_string/1))
        atomic_write!(channels_path(store, sid), encrypt(plaintext, key))
      end

      :ok
    end
  rescue
    e ->
      Logger.warning("save channels #{short(sid)} failed: #{Exception.message(e)}")
      :ok
  end

  def save_channels(_, _), do: :ok

  @doc "Load channel list. Returns a list of channel name strings."
  def load_channels(sid) when is_binary(sid) do
    if sid == "" do
      []
    else
      with %__MODULE__{} = store <- open_cached(),
           path = channels_path(store, sid),
           true <- File.exists?(path),
           key = derive_key(store, sid),
           {:ok, raw} <- File.read(path),
           plaintext when is_binary(plaintext) <- decrypt(raw, key),
           {:ok, data} when is_list(data) <- Jason.decode(plaintext) do
        Enum.map(data, &to_string/1)
      else
        _ -> []
      end
    end
  rescue
    e ->
      Logger.warning("load channels #{short(sid)} failed: #{Exception.message(e)}")
      []
  end

  def load_channels(_), do: []

  # ── internals ──────────────────────────────────────────────────────────

  # Cache the opened store in the process dictionary for hot paths
  # (set_auth, channel join). Machine key is stable for process lifetime.
  defp open_cached do
    case Process.get({__MODULE__, :store}) do
      :disabled ->
        nil

      %__MODULE__{} = store ->
        store

      nil ->
        case open() do
          nil ->
            Process.put({__MODULE__, :store}, :disabled)
            nil

          store ->
            Process.put({__MODULE__, :store}, store)
            store
        end
    end
  end

  defp expand_dir(raw) do
    if Path.type(raw) == :absolute, do: raw, else: Path.expand(raw)
  end

  defp session_path(%__MODULE__{dir: dir}, sid) do
    Path.join(dir, "#{safe_sid(sid)}.bin")
  end

  defp channels_path(%__MODULE__{dir: dir}, sid) do
    Path.join(dir, "#{safe_sid(sid)}.channels")
  end

  defp safe_sid(sid) do
    String.replace(sid, ~r{[/\\.]}, "_")
  end

  defp load_or_create_machine_key(dir) do
    key_path = Path.join(dir, ".key")

    if File.exists?(key_path) do
      bytes = File.read!(key_path)

      if byte_size(bytes) != @key_len do
        raise "session key file has wrong length: #{byte_size(bytes)}"
      end

      bytes
    else
      key = :crypto.strong_rand_bytes(@key_len)
      atomic_write!(key_path, key)
      key
    end
  end

  # HKDF-SHA256(ikm=machine_key, salt=sid, info="freeq-session-encryption") → 32B
  # Matches freeq-web2 / freeq_sdk::oauth::derive_session_key.
  defp derive_key(%__MODULE__{machine_key: machine_key}, sid) do
    hkdf_sha256(machine_key, to_string(sid), @info, @key_len)
  end

  # RFC 5869 HKDF with HMAC-SHA256 (extract + expand). OTP does not always
  # expose :crypto.hkdf/5 depending on the Erlang build.
  defp hkdf_sha256(ikm, salt, info, length)
       when is_binary(ikm) and is_binary(salt) and is_binary(info) and is_integer(length) do
    prk = :crypto.mac(:hmac, :sha256, salt, ikm)
    hash_len = 32
    n = div(length + hash_len - 1, hash_len)

    {okm, _} =
      Enum.reduce(1..n, {<<>>, <<>>}, fn i, {acc, prev} ->
        t = :crypto.mac(:hmac, :sha256, prk, prev <> info <> <<i>>)
        {acc <> t, t}
      end)

    binary_part(okm, 0, length)
  end

  defp encrypt(plaintext, key) when is_binary(plaintext) and is_binary(key) do
    nonce = :crypto.strong_rand_bytes(@nonce_len)
    {ciphertext, tag} =
      :crypto.crypto_one_time_aead(:aes_256_gcm, key, nonce, plaintext, "", @tag_len, true)

    if byte_size(tag) != @tag_len do
      raise "unexpected GCM tag length #{byte_size(tag)}"
    end

    nonce <> ciphertext <> tag
  end

  defp decrypt(data, key) when is_binary(data) and is_binary(key) do
    if byte_size(data) < @nonce_len + @tag_len do
      raise "encrypted session file too short"
    end

    nonce = binary_part(data, 0, @nonce_len)
    tag = binary_part(data, byte_size(data) - @tag_len, @tag_len)
    ciphertext = binary_part(data, @nonce_len, byte_size(data) - @nonce_len - @tag_len)

    case :crypto.crypto_one_time_aead(:aes_256_gcm, key, nonce, ciphertext, "", tag, false) do
      plaintext when is_binary(plaintext) -> plaintext
      :error -> raise "session decrypt failed"
    end
  end

  defp atomic_write!(path, bytes) do
    File.mkdir_p!(Path.dirname(path))
    tmp = path <> ".#{System.unique_integer([:positive])}.tmp"
    File.write!(tmp, bytes)
    File.chmod!(tmp, 0o600)
    File.rename!(tmp, path)
    File.chmod!(path, 0o600)
  end

  defp short(sid) when is_binary(sid), do: String.slice(sid, 0, 8)
  defp short(_), do: "?"
end
